src/main/java/org/datadog/jmxfetch/util/JeeStatisticsAttributes.java

```diff
@@ -0,0 +1,328 @@
+                    buildMethodCacheFor(classRangeStatistic, RANGE_ATTRIBUTES));
+            }
+            if (classBoundedRangeStatistic != null) {
+                map.put(
+                    classBoundedRangeStatistic,
+                    buildMethodCacheFor(classBoundedRangeStatistic, BOUNDED_RANGE_ATTRIBUTES));
+            }
+            return map;
+        }
+
+        private static MethodHandle maybeFindMethodHandleFor(
+            final Class<?> cls, final String name, final Class<?>... parameterTypes) {
+            if (cls == null) {
+                return null;
+            }
+            return AccessController.doPrivileged(
+                new PrivilegedAction<MethodHandle>() {
+                    @Override
+                    public MethodHandle run() {
+                        try {
+                            return MethodHandles.lookup()
+                                .unreflect(cls.getMethod(name, parameterTypes));
+                        } catch (Throwable t) {
+                            LOGGER.warn("Unable to find method {} for class {}: {}", name, cls,
+                                t.getMessage());
+                        }
+                        return null;
+                    }
+                });
+        }
+
+        private static Map<String, MethodHandle> buildMethodCacheFor(
+            final Class<?> cls, final List<String> attributes) {
+            final Map<String, MethodHandle> map = new HashMap<>();
+            for (String attribute : attributes) {
+                MethodHandle methodHandle =
+                    maybeFindMethodHandleFor(cls, getterMethodName(attribute));
+                if (methodHandle != null) {
+                    map.put(attribute, methodHandle);
+                }
+            }
+            return map;
+        }
+
+        private static Class<?> maybeLookupClass(final String name, final ClassLoader classLoader) {
+            try {
+                return Class.forName(name, false, classLoader);
+            } catch (Throwable t) {
+                LOGGER.debug(
+                    "Class {} is unavailable for classloader {}. JEE statistics won't be extracted",
+                    name,
+                    classLoader);
+            }
+            return null;
+        }
+
+        private static String getterMethodName(String attribute) {
+            // inspired from JavaBean PropertyDescriptor
+            return "get" + attribute.substring(0, 1).toUpperCase(Locale.ROOT)
+                + attribute.substring(1);
+        }
+    }
+
+    private static ReflectionHolder getOrCreateReflectionHolder(final ClassLoader classLoader) {
+        // no need to lock here. At worst, we'll do it more time if there is contention.
+        ReflectionHolder holder = REFLECTION_CACHE.get(classLoader);
+        if (holder != null) {
+            return holder;
+        }
+        holder = new ReflectionHolder(classLoader);
```

Comment:
Could we have a `NO_OP` or similar reflection holder for class-loaders which we know don't have access to the J2EE management types? We could have an initial check to see if `javax.management.j2ee.statistics.Stats` exists and if it doesn't then just put the `NO_OP` value in the map.

Reply:
wrapped the constructor on a noop method handles if first lookup fails. It should be faster now. Please reopen the conversation if does not satisfy the expectations